enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,532 @@
+  on GCP, Azure, AWS, and OpenStack platforms, other platforms, such as IBM or Alibaba do not engage CNCC for the moment.
+
+- There are also several platform specific hacks,
+  like the [access restriction to a metadata service](https://github.com/openshift/ovn-kubernetes/blob/8d5db0bae387de423eb30db75142b0ad87df4f3f/go-controller/pkg/cni/OCP_HACKS.go#L18),
+  but it is not entirely connected with the operator itself and is more CNI plugin specific.
+
+For the [phase 1](#implementation-phases) of this project, there seems to be no particular action or API knobs needed regarding
+the addition of the "External" platform type. Just ensure that CNO is non-reactive to the "External" platform type
+and behaves the same as in the "None" platform case.
+In the future, we possibly might want to make CNO more tolerant of partner CNCC implementations
+and design a way for configuring platform-specific CNI behaviour.
+
+### Workflow Description
+
+During [phase 1](#implementation-phases) proposed changes intended to have no effect and the "External"
+platform type should be handled the same as the "None", so no specific user interaction is expected.
+
+### API Extensions
+
+A new optional constant of `PlatformType` type will be added to the "openshift/api".
+
+```go
+const (
+    ...
+    // ExternalPlatformType represent generic infrastructure provider. Provider-specific components should be supplemented separately.
+    ExternalPlatformType PlatformType = "External"
+    ...
+)
+```
+
+Additionally, the respective external platform spec and status should be added to the infrastructure resource.
+
+```go
+type CloudControllerManagerMode string
+
+const (
+    // Cloud Controller Manager is enabled and expected to be supplied.
+    // Signaling MCO to set `--cloud-provider=external` flag to the kubelets.
+    CloudControllerManagerEnabled CloudControllerManagerMode = "Enabled"
+    // No Cloud Controller Manager is expected to be supplied.
+    // Signaling MCO not to set `--cloud-provider` flag to the kubelets.
+    CloudControllerManagerDisabled CloudControllerManagerMode = "Disabled"
+)
+
+type CloudControllerManagerSettings struct {
+    // Mode determines whether or not an external Cloud Controller Manager is expected to 
+    // be presented in the cluster.
+    // For engaging an external Cloud Controller Manager, certain flags are expected to be set to the kubelets.
+    // https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+    //
+    // When enabled, the respective operator (machine config operator) should set `--cloud-provider=external` flag to the kubelet.
+    // When omitted or disabled, no `cloud-provider` flag should be set.
+    // +kubebuilder:validation:Enum=Enabled;Disabled
+    // +optional
+    Mode CloudControllerManagerMode `json:"mode,omitempty"`
```

Comment:
I want to have a think about the naming here and the values, we can have a discussion about it on the API PR, general structure is ok though

I'm thinking if we want `state: External | None`, because we aren't enabling something vs disabling it, we are switching on the presence of something set up by someone else, WDYT? Not 100% on this, need to have a deeper think. 

Here `None` means I don't have a provider and `External` means I have one and I'm installing it, this could mean down the line we add a `Managed` option where the user provides some additional config and CCMO manages it 🤔

Reply:
i think you've parsed the underlying concerns here Joel. we definitely want to use the `External` option for now to signal that the MCO needs to deploy kubelets with the `--cloud-provider=external` flag enabled.

it's possible we could expand this later, especially if we do more to have CCMO manage the external CCM, but i think the core functionality here is still about ensuring that the MCO makes the kubelet systemd units properly from bootstrap forward.

Reply:
I do like the idea with `state: External | None` and further notion of introducing `Managed` option. Will update the document.